File: revenue/hive/caption-transcript-intake/workbench.js
'use strict';
const $ = (id) => document.getElementById(id);
const state = {rows: [], active: null, busy: false, next: 1};
const limits = {max_files: 20, max_file_bytes: 2000000, max_batch_bytes: 20000000};

function node(tag, text, cls) {
  const result = document.createElement(tag);
  if (text !== undefined) result.textContent = text;
  if (cls) result.className = cls;
  return result;
}
function status(text, error = false) {
  $('global-status').textContent = text;
  $('global-status').className = error ? 'message error' : 'message';
}
function clock(ms) {
  const hours = Math.floor(ms / 3600000);
  const minutes = Math.floor(ms / 60000) % 60;
  const seconds = Math.floor(ms / 1000) % 60;
  return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}.${String(ms % 1000).padStart(3, '0')}`;
}
function active() { return state.rows.find(row => row.id === state.active); }
function selected() { return state.rows.filter(row => row.selected); }
function summaryLabel(row) {
  if (!row.result) return 'Not previewed / options changed';
  if (row.result.ok) return `${row.result.cue_count} cues · ready to export`;
  return row.result.parsed ? 'Parsed · export needs attention' : 'Input needs attention';
}
function renderList() {
  $('file-list').replaceChildren();
  for (const row of state.rows) {
    const li = node('li', undefined, `file-row${row.id === state.active ? ' active' : ''}`);
    const check = document.createElement('input');
    check.type = 'checkbox'; check.checked = row.selected;
    check.setAttribute('aria-label', `Include ${row.file.name} in batch`);
    check.addEventListener('change', () => { row.selected = check.checked; renderList(); });
    const pick = node('button', undefined, 'pick'); pick.type = 'button';
    pick.append(node('span', row.file.name), node('small', summaryLabel(row), row.result && !row.result.ok ? 'bad' : ''));
    pick.addEventListener('click', () => { state.active = row.id; renderList(); renderDetail(); });
    const remove = node('button', '×', 'remove'); remove.type = 'button';
    remove.setAttribute('aria-label', `Remove ${row.file.name}`);
    remove.addEventListener('click', () => {
      state.rows = state.rows.filter(item => item !== row);
      if (state.active === row.id) state.active = state.rows[0]?.id ?? null;
      renderList(); renderDetail();
    });
    li.append(check, pick, remove); $('file-list').append(li);
  }
  $('empty-list').hidden = state.rows.length > 0;
  $('selection-count').textContent = `${selected().length} of ${state.rows.length} files selected`;
  $('preview-all').disabled = selected().length === 0;
  $('export-batch').disabled = selected().length === 0;
  $('clear').disabled = state.rows.length === 0;
}
function renderDetail() {
  const row = active();
  $('empty-detail').hidden = !!row; $('detail').hidden = !row;
  if (!row) return;
  $('file-name').textContent = row.file.name;
  $('file-size').textContent = `${row.file.size.toLocaleString()} original bytes · original file is not edited`;
  for (const field of ['title', 'format', 'encoding', 'speaker', 'duration']) $(field).value = row[field];
  $('synthetic').checked = row.synthetic;
  $('query').value = row.query;
  renderResult();
}
function renderResult() {
  const row = active(); if (!row) return;
  const result = row.result;
  $('summary').hidden = !result?.parsed;
  $('preview').hidden = !result?.parsed;
  $('file-status').textContent = result ? (result.ok ? `Ready: ${result.episode_import ? 'generic handoff + episode import' : 'generic handoff'}. Recording not verified.` : result.error) : 'Options are ready to preview. Exports always parse the current options again.';
  $('file-status').className = result && !result.ok ? 'message error' : 'message';
  if (!result?.parsed) return;
  $('metrics').replaceChildren();
  for (const [value, label] of [[result.cue_count, 'caption cues'], [result.speaker_count, 'supplied speaker labels'], [result.overlapping_cues, 'cues overlapping earlier cues']]) {
    const metric = node('div', undefined, 'metric');
    metric.append(node('strong', value), node('span', label)); $('metrics').append(metric);
  }
  $('source-info').replaceChildren(
    node('p', `Caption span ${clock(result.first_caption_ms)} – ${clock(result.last_caption_ms)}. This is not the recording duration.`),
    node('p', `${result.format.toUpperCase()} · ${result.encoding} · ${result.source_bytes.toLocaleString()} original bytes`),
    node('p', `Speaker labels: ${result.speakers.map(item => `${item.name} (${item.cues})`).join(', ')}${result.speaker_count > 100 ? ' · first 100 labels shown' : ''}`),
    node('p', 'Original SHA-256:'), node('code', result.source_sha256));
  const page = result.preview;
  $('cues').replaceChildren();
  for (const cue of page.cues) {
    const article = node('article', undefined, 'cue');
    article.append(node('h3', `${cue.id} · ${cue.speaker}`),
      node('small', `${clock(cue.start_ms)} – ${clock(cue.end_ms)} · source line ${cue.line}`, 'quiet'),
      node('p', cue.text));
    if (cue.text_truncated) article.append(node('small', 'Preview shortened after 4,000 characters; export retains the complete text.', 'quiet'));
    const raw = document.createElement('details');
    raw.append(node('summary', `Original cue ${cue.source_id ?? '(no source ID)'} · timing and markup`),
      node('pre', `${cue.timing}\n${cue.raw_payload}${cue.raw_truncated ? '\n[Preview shortened; original is retained in the export.]' : ''}`));
    article.append(raw); $('cues').append(article);
  }
  if (!page.cues.length) $('cues').append(node('p', 'No cues match this search.', 'quiet'));
  $('page-label').textContent = page.matched ? `${page.offset + 1}–${page.offset + page.cues.length} of ${page.matched} matching cues` : '0 matching cues';
  $('previous').disabled = page.offset === 0;
  $('next').disabled = page.offset + page.page_size >= page.matched;
}
async function busy(task) {
  if (state.busy) return;
  state.busy = true; $('workspace').disabled = true;
  try { await task(); } catch (error) { status(error.message || 'The operation failed.', true); }
  finally { state.busy = false; $('workspace').disabled = false; }
}
function asBase64(file) {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onerror = () => reject(new Error(`Cannot read ${file.name}. Select it again.`));
    reader.onload = () => resolve(String(reader.result).split(',')[1]);
    reader.readAsDataURL(file);
  });
}
async function addFiles(files) {
  await busy(async () => {
    const incoming = Array.from(files);
    if (incoming.length + state.rows.length > limits.max_files) throw new Error(`Keep at most ${limits.max_files} files in a batch.`);
    if (incoming.some(file => file.size === 0 || file.size > limits.max_file_bytes)) throw new Error('Each file must contain 1–2,000,000 bytes. No files from this selection were added.');
    const bytes = [...state.rows.map(row => row.file), ...incoming].reduce((total, file) => total + file.size, 0);
    if (bytes > limits.max_batch_bytes) throw new Error('Original files exceed the 20 MB batch limit. Remove files or split the batch.');
    // Add atomically after every file has been read; a failed read cannot create a partial selection.
    const additions = [];
    for (const file of incoming) {
      const extension = file.name.split('.').pop().toLowerCase();
      additions.push({id: state.next++, file, source: await asBase64(file), selected: true,
        title: file.name.replace(/\.(srt|vtt)$/i, ''), format: ['srt', 'vtt'].includes(extension) ? extension : '',
        encoding: 'utf-8-sig', speaker: 'Unspecified', duration: '', synthetic: false, query: '', result: null});
    }
    state.rows.push(...additions); state.active ??= additions[0]?.id ?? null;
    renderList(); renderDetail(); status(`Added ${additions.length} file(s). Choose a file to edit options, then preview the selected batch.`);
  });
  $('files').value = '';
}
function requestRow(row, offset = 0) {
  return {name: row.file.name, source_base64: row.source, title: row.title, format: row.format,
    encoding: row.encoding, speaker: row.speaker, duration_seconds: row.duration || null,
    synthetic_demo: row.synthetic, query: row.query, offset};
}
function applyResults(rows, result) {
  if (Array.isArray(result.results)) {
    result.results.forEach((item, index) => { if (rows[index]) rows[index].result = item; });
    renderList(); renderResult();
  }
}
async function convert(rows, endpoint, offset = 0) {
  if (!rows.length) return;
  await busy(async () => {
    status(endpoint === 'preview' ? `Reading ${rows.length} file(s)…` : `Building ${rows.length} complete handoff(s)…`);
    const response = await fetch(`/api/${endpoint}`, {method: 'POST', headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({files: rows.map(row => requestRow(row, offset))})});
    if ((response.headers.get('content-type') || '').includes('application/json')) {
      const result = await response.json(); applyResults(rows, result);
      if (!response.ok || result.error) throw new Error(result.error || 'Nothing downloaded. Fix or deselect the files marked for attention.');
      const ready = (result.results || []).filter(item => item.ok).length;
      status(`${ready} of ${rows.length} file(s) ready to export.${ready < rows.length ? ' Inspect the marked file(s); no captions were dropped.' : ' Words and recording still require human review.'}`, ready < rows.length);
      return;
    }
    if (!response.ok) throw new Error(`Local server returned HTTP ${response.status}; no completed download.`);
    const blob = await response.blob();
    const url = URL.createObjectURL(blob);
    const link = document.createElement('a'); link.href = url;
    const stem = rows[0].file.name.replace(/\.(srt|vtt)$/i, '').replace(/[^a-zA-Z0-9_-]+/g, '-').replace(/^-+|-+$/g, '').slice(0, 60) || 'captions';
    link.download = endpoint === 'batch' ? 'caption-batch.zip' : `${stem}-handoff.zip`;
    document.body.append(link); link.click(); link.remove();
    setTimeout(() => URL.revokeObjectURL(url), 60000);
    status(`Download prepared for ${rows.length} file(s). Check your browser downloads. Originals and complete captions are retained; no application import was performed.`);
  });
}
for (const field of ['title', 'format', 'encoding', 'speaker', 'duration']) {
  $(field).addEventListener('input', () => {
    const row = active(); if (!row) return;
    row[field] = $(field).value; row.result = null; renderList(); renderResult();
  });
}
$('synthetic').addEventListener('change', () => {
  const row = active(); if (!row) return;
  row.synthetic = $('synthetic').checked; row.result = null; renderList(); renderResult();
});
$('files').addEventListener('change', () => addFiles($('files').files));
$('drop').addEventListener('dragover', event => { event.preventDefault(); if (!state.busy) $('drop').classList.add('drag'); });
$('drop').addEventListener('dragleave', () => $('drop').classList.remove('drag'));
$('drop').addEventListener('drop', event => { event.preventDefault(); $('drop').classList.remove('drag'); if (!state.busy) addFiles(event.dataTransfer.files); });
$('clear').addEventListener('click', () => { state.rows = []; state.active = null; renderList(); renderDetail(); status('Selection cleared. Downloaded files are unchanged.'); });
$('preview-all').addEventListener('click', () => convert(selected(), 'preview'));
$('export-batch').addEventListener('click', () => convert(selected(), 'batch'));
$('preview-one').addEventListener('click', () => { const row = active(); if (row) convert([row], 'preview'); });
$('export-one').addEventListener('click', () => { const row = active(); if (row) convert([row], 'export'); });
function search() { const row = active(); if (row) { row.query = $('query').value; convert([row], 'preview'); } }
$('search').addEventListener('click', search);
$('query').addEventListener('keydown', event => { if (event.key === 'Enter') { event.preventDefault(); search(); } });
$('previous').addEventListener('click', () => { const row = active(); if (row?.result?.preview) convert([row], 'preview', Math.max(0, row.result.preview.offset - row.result.preview.page_size)); });
$('next').addEventListener('click', () => { const row = active(); if (row?.result?.preview) convert([row], 'preview', row.result.preview.offset + row.result.preview.page_size); });
renderList(); renderDetail();
fetch('/api/info').then(response => { if (!response.ok) throw new Error('Local server unavailable.'); return response.json(); })
  .then(info => { Object.assign(limits, info); status('Ready. Choose supplied caption files; nothing is sent to an external service.'); })
  .catch(error => status(`${error.message} Start workbench.py and open its local address.`, true));

File: revenue/hive/caption-transcript-intake/workbench.py
#!/usr/bin/env python3
"""Local batch caption workbench over the existing source-preserving parser."""
from __future__ import annotations

import argparse
import base64
import binascii
from collections import Counter
import hashlib
from http.server import BaseHTTPRequestHandler, HTTPServer
import io
import json
from pathlib import Path
import re
import sys
import traceback
from urllib.parse import urlsplit
import zipfile

from caption_intake import (MAX_BYTES, MAX_CUES, IntakeError, canonical_episode,
                            create_bundle, json_bytes, parse_captions)

MAX_FILES = 20
MAX_BATCH_BYTES = 20_000_000
MAX_REQUEST_BYTES = 28_000_000  # Includes base64 expansion and request metadata.
MAX_BATCH_CUES = 50_000
PAGE_SIZE = 100
PREVIEW_CHARS = 4_000
ROOT = Path(__file__).resolve().parent


def text_field(row: dict, key: str, default: str, limit: int) -> str:
    value = row.get(key, default)
    if not isinstance(value, str) or not value.strip() or len(value) > limit:
        raise IntakeError(f'{key} must contain 1..{limit} characters')
    if any(ord(char) < 32 or ord(char) == 127 for char in value):
        raise IntakeError(f'{key} contains control characters')
    return value.strip()


def read_upload(row: object) -> tuple[str, bytes, dict]:
    if not isinstance(row, dict):
        raise IntakeError('each file must be an object')
    name = text_field(row, 'name', 'captions', 500)
    encoded = row.get('source_base64')
    if not isinstance(encoded, str) or not 0 < len(encoded) <= 4 * ((MAX_BYTES + 2) // 3):
        raise IntakeError(f'source must contain 1..{MAX_BYTES} bytes')
    try:
        source = base64.b64decode(encoded, validate=True)
    except (ValueError, binascii.Error) as exc:
        raise IntakeError('source_base64 must encode the original caption bytes') from exc
    if not 0 < len(source) <= MAX_BYTES:
        raise IntakeError(f'source must contain 1..{MAX_BYTES} bytes')
    settings = {
        'fmt': text_field(row, 'format', Path(name).suffix.lower().lstrip('.'), 8),
        'title': text_field(row, 'title', Path(name).stem, 300),
        'encoding': text_field(row, 'encoding', 'utf-8-sig', 80),
        'default_speaker': text_field(row, 'speaker', 'Unspecified', 200),
    }
    return name, source, settings


def describe(parsed: dict, row: dict) -> dict:
    segments = parsed['segments']
    provenance = parsed['provenance']
    overlaps, end = 0, 0
    for cue in segments:
        overlaps += cue['start_ms'] < end
        end = max(end, cue['end_ms'])
    voices = Counter(cue['speaker'] for cue in segments)
    query = row.get('query', '')
    offset = row.get('offset', 0)
    if not isinstance(query, str) or len(query) > 256:
        raise IntakeError('preview query must be text of at most 256 characters')
    if type(offset) is not int or not 0 <= offset <= MAX_CUES:
        raise IntakeError(f'preview offset must be an integer from 0 to {MAX_CUES}')
    needle = query.casefold()
    indices = [i for i, cue in enumerate(segments)
               if not needle or any(needle in str(value).casefold() for value in
                                    (cue['id'], cue['text'], cue['speaker'],
                                     provenance['cues'][i]['source_id']))]
    offset = min(offset, max(0, ((len(indices) - 1) // PAGE_SIZE) * PAGE_SIZE))
    preview = []
    for i in indices[offset:offset + PAGE_SIZE]:
        cue, original = segments[i], provenance['cues'][i]
        preview.append({**cue, 'text': cue['text'][:PREVIEW_CHARS],
                        'text_truncated': len(cue['text']) > PREVIEW_CHARS,
                        'source_id': original['source_id'], 'line': original['line'],
                        'timing': original['timing'], 'settings': original['settings'],
                        'raw_payload': original['raw_payload'][:PREVIEW_CHARS],
                        'raw_truncated': len(original['raw_payload']) > PREVIEW_CHARS})
    return {
        'title': parsed['title'], 'cue_count': len(segments),
        'first_caption_ms': segments[0]['start_ms'], 'last_caption_ms': end,
        'overlapping_cues': overlaps, 'speaker_count': len(voices),
        'speakers': [{'name': name, 'cues': count} for name, count in voices.most_common(100)],
        'source_sha256': provenance['source_sha256'], 'source_bytes': provenance['source_bytes'],
        'format': provenance['format'], 'encoding': provenance['encoding'],
        'media_verified': False,
        'preview': {'offset': offset, 'page_size': PAGE_SIZE, 'matched': len(indices),
                    'query': query, 'cues': preview},
    }


def process_files(document: object, *, build: bool = False) -> tuple[list[dict], list[tuple[str, bytes]]]:
    if not isinstance(document, dict) or not isinstance(document.get('files'), list):
        raise IntakeError('request must contain a files array')
    files = document['files']
    if not 1 <= len(files) <= MAX_FILES:
        raise IntakeError(f'select 1..{MAX_FILES} files')
    results: list[dict] = []
    bundles: list[tuple[str, bytes]] = []
    source_total = cue_total = 0
    for index, row in enumerate(files):
        result = {'index': index, 'name': f'File {index + 1}', 'ok': False, 'parsed': False}
        results.append(result)
        try:
            name, source, settings = read_upload(row)
            result['name'] = name
            source_total += len(source)
            if source_total > MAX_BATCH_BYTES:
                raise IntakeError(f'selected originals exceed {MAX_BATCH_BYTES} bytes; split the batch')
            parsed = parse_captions(source, **settings)
            cue_total += len(parsed['segments'])
            if cue_total > MAX_BATCH_CUES:
                raise IntakeError(f'selected captions exceed {MAX_BATCH_CUES} cues; split the batch')
            result.update(describe(parsed, row))
            result['parsed'] = True
            duration = row.get('duration_seconds')
            if duration == '':
                duration = None
            synthetic = row.get('synthetic_demo', False)
            if type(synthetic) is not bool:
                raise IntakeError('synthetic_demo must be a boolean')
            result['episode_import'] = duration is not None
            if build:
                payload = create_bundle(source, parsed, duration_seconds=duration, synthetic_demo=synthetic)
                stem = re.sub(r'[^a-zA-Z0-9_-]+', '-', Path(name).stem).strip('-')[:60] or 'captions'
                member = f'{index + 1:02d}-{stem}.zip'
                bundles.append((member, payload))
                result['bundle'] = {'path': member, 'bytes': len(payload),
                                    'sha256': hashlib.sha256(payload).hexdigest()}
            elif duration is not None:
                canonical_episode(parsed, duration, synthetic_demo=synthetic)
            result['ok'] = True
        except IntakeError as exc:
            result['error'] = str(exc)
    return results, bundles


def batch_archive(results: list[dict], bundles: list[tuple[str, bytes]]) -> bytes:
    """Wrap the existing handoff ZIPs; no partial export or path from an upload."""
    if not results or not all(row['ok'] for row in results) or len(bundles) != len(results):
        raise IntakeError('fix or deselect every invalid file before exporting this batch')
    manifest = {'schema': 'caption-workbench-batch/v1', 'complete': True,
                'media_verified': False, 'files': [
                    {key: row[key] for key in ('name', 'title', 'source_sha256', 'source_bytes',
                                               'format', 'encoding', 'cue_count', 'episode_import', 'bundle')}
                    for row in results]}
    output = io.BytesIO()
    with zipfile.ZipFile(output, 'w', compression=zipfile.ZIP_STORED) as archive:
        for name, payload in [*bundles, ('batch-manifest.json', json_bytes(manifest))]:
            archive.writestr(zipfile.ZipInfo(name, (1980, 1, 1, 0, 0, 0)), payload)
    return output.getvalue()


def reject_constant(value: str) -> None:
    raise IntakeError(f'JSON constant {value} is not supported')


class WorkbenchServer(HTTPServer):
    """One request at a time bounds concurrent in-memory conversion work."""
    def __init__(self, port: int):
        self.assets = {
            '/': ((ROOT / 'workbench.html').read_bytes(), 'text/html; charset=utf-8'),
            '/workbench.js': ((ROOT / 'workbench.js').read_bytes(), 'text/javascript; charset=utf-8'),
        }
        super().__init__(('127.0.0.1', port), WorkbenchHandler)


class WorkbenchHandler(BaseHTTPRequestHandler):
    server: WorkbenchServer

    def setup(self) -> None:
        super().setup()
        self.connection.settimeout(30)

    def log_message(self, format: str, *args: object) -> None:
        # Do not echo uploaded names, captions, query strings, or request headers.
        pass

    def reply(self, status: int, payload: bytes, kind: str = 'application/json; charset=utf-8',
              filename: str | None = None) -> None:
        self.send_response(status)
        self.send_header('Content-Type', kind)
        self.send_header('Content-Length', str(len(payload)))
        self.send_header('Cache-Control', 'no-store')
        self.send_header('X-Content-Type-Options', 'nosniff')
        self.send_header('Content-Security-Policy', "default-src 'none'; script-src 'self'; "
                         "style-src 'unsafe-inline'; connect-src 'self'; base-uri 'none'; frame-ancestors 'none'")
        if filename:
            self.send_header('Content-Disposition', f'attachment; filename="{filename}"')
        self.end_headers()
        self.wfile.write(payload)

    def do_GET(self) -> None:
        path = urlsplit(self.path).path
        if path in self.server.assets:
            payload, kind = self.server.assets[path]
            self.reply(200, payload, kind)
        elif path == '/api/info':
            self.reply(200, json_bytes({'service': 'caption-workbench', 'max_files': MAX_FILES,
                                       'max_file_bytes': MAX_BYTES, 'max_batch_bytes': MAX_BATCH_BYTES,
                                       'max_batch_cues': MAX_BATCH_CUES, 'page_size': PAGE_SIZE}))
        elif path == '/favicon.ico':
            self.reply(204, b'')
        else:
            self.reply(404, json_bytes({'error': 'route not found'}))

    def read_document(self) -> object:
        if self.headers.get('Transfer-Encoding'):
            raise IntakeError('send a Content-Length JSON request; chunked bodies are not supported')
        if self.headers.get('Content-Type', '').split(';')[0].strip().lower() != 'application/json':
            raise IntakeError('Content-Type must be application/json')
        lengths = self.headers.get_all('Content-Length', [])
        if len(lengths) != 1 or not lengths[0].isdigit():
            raise IntakeError('one integer Content-Length is required')
        length = int(lengths[0])
        if not 0 < length <= MAX_REQUEST_BYTES:
            raise IntakeError(f'JSON request exceeds {MAX_REQUEST_BYTES} bytes or is empty')
        payload = self.rfile.read(length)
        if len(payload) != length:
            raise IntakeError('incomplete request; original files were not stored')
        try:
            return json.loads(payload.decode('utf-8'), parse_constant=reject_constant)
        except (ValueError, UnicodeError, RecursionError) as exc:
            raise IntakeError('invalid UTF-8 JSON request') from exc

    def do_POST(self) -> None:
        path = urlsplit(self.path).path
        if path not in ('/api/preview', '/api/export', '/api/batch'):
            self.reply(404, json_bytes({'error': 'route not found'}))
            return
        try:
            document = self.read_document()
            if path == '/api/export' and (not isinstance(document, dict)
                                         or not isinstance(document.get('files'), list)
                                         or len(document['files']) != 1):
                raise IntakeError('individual export requires exactly one file')
            results, bundles = process_files(document, build=path != '/api/preview')
            ok = all(row['ok'] for row in results)
            if path == '/api/preview' or not ok:
                self.reply(200 if path == '/api/preview' else 422,
                           json_bytes({'ok': ok, 'results': results}))
            elif path == '/api/export':
                self.reply(200, bundles[0][1], 'application/zip', bundles[0][0])
            else:
                self.reply(200, batch_archive(results, bundles), 'application/zip', 'caption-batch.zip')
        except IntakeError as exc:
            self.reply(400, json_bytes({'error': str(exc)}))
        except (TimeoutError, ConnectionError):
            self.close_connection = True
        except Exception:
            # Genuine programming failures stay visible; never turn them into a successful export.
            traceback.print_exc(file=sys.stderr)
            self.reply(500, json_bytes({'error': 'conversion failed unexpectedly; see the server terminal'}))


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--port', type=int, default=8766, help='loopback port (0 selects an available port)')
    args = parser.parse_args(argv)
    if not 0 <= args.port <= 65535:
        parser.error('--port must be from 0 to 65535')
    try:
        with WorkbenchServer(args.port) as server:
            print(f'Caption workbench: http://127.0.0.1:{server.server_port}/', flush=True)
            print('Originals are processed in memory. Close the tab to clear its selection; Ctrl+C stops the server.', flush=True)
            server.serve_forever()
    except KeyboardInterrupt:
        return 0
    except OSError as exc:
        print(f'caption-workbench: {exc}', file=sys.stderr)
        return 2
    return 0


if __name__ == '__main__':
    raise SystemExit(main())
